File: Solutions/1679. Max Number of K-Sum Pairs/1679. Max Number of K-Sum Pairs.go
// O(NlogN)
func maxOperations(nums []int, k int) int {
    sort.Ints(nums)
    res:=0
    i,j:=0,len(nums)-1
    for i<j{
        if nums[i]+nums[j]==k{
            res,i,j=res+1,i+1,j-1
        }else if nums[i]+nums[j]>k{
            j-=1
        }else{
            i+=1
        }
    }
    return res
}

// O(N)
func maxOperations(nums []int, k int) int {
    um:=make(map[int]int)
    for _,n:=range nums{
        um[n]+=1
    }
    res:=0
    for n,cnt:=range um{
        if k-n==n{
            res+=cnt/2
        }else if val,ok:=um[k-n];ok{
            tmp:=min(cnt,val)
            res+=tmp
            um[n]-=tmp
            um[k-n]-=tmp
        }
    }
    return res
}

func min(i,j int) int{
    if i<j{
        return i
    }
    return j
}

File: Solutions/1679. Max Number of K-Sum Pairs/1679. Max Number of K-Sum Pairs.cpp
// O(NlogN)
class Solution {
public:
    int maxOperations(vector<int>& nums, int k) {
        sort(nums.begin(),nums.end());
        int res=0;
        int i=0,j=nums.size()-1;
        while(i<j){
            if(nums[i]+nums[j]==k){
                ++res,++i,--j;
            }else if(nums[i]+nums[j]>k){
                --j;
            }else{
                ++i;
            }
        }
        return res;
    }
};

// O(N)
class Solution {
public:
    int maxOperations(vector<int>& nums, int k) {
        unordered_map<int,int> um;
        for(auto& n:nums){
            ++um[n];
        }
        int res=0;
        for(auto& [n,cnt]:um){
            if(k-n==n){
                res+=cnt/2;
            }else if(um.count(k-n)){
                int tmp=min(cnt,um[k-n]);
                res+=tmp;
                um[n]-=tmp;
                um[k-n]=tmp;
            }
        }
        return res;
    }
};